File: binary-search/4-Median-of-Two-Sorted-Arrays.go
func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
\tif len(nums1) > len(nums2) {
\t\tnums1, nums2 = nums2, nums1
\t}

\tm, n := len(nums1), len(nums2)
\tlow, high := 0, m

\tfor low <= high {
\t\tpartitionX := (low + high) / 2
\t\tpartitionY := (m + n + 1) / 2 - partitionX

\t\tmaxX := math.MinInt64
\t\tif partitionX > 0 {
\t\t\tmaxX = nums1[partitionX-1]
\t\t}

\t\tminX := math.MaxInt64
\t\tif partitionX < m {
\t\t\tminX = nums1[partitionX]
\t\t}

\t\tmaxY := math.MinInt64
\t\tif partitionY > 0 {
\t\t\tmaxY = nums2[partitionY-1]
\t\t}

\t\tminY := math.MaxInt64
\t\tif partitionY < n {
\t\t\tminY = nums2[partitionY]
\t\t}

\t\tif maxX <= minY && maxY <= minX {
\t\t\tif (m+n)%2 == 0 {
\t\t\t\treturn (float64(max(maxX, maxY)) + float64(min(minX, minY))) / 2.0
\t\t\t}
\t\t\treturn float64(max(maxX, maxY))
\t\t} else if maxX > minY {
\t\t\thigh = partitionX - 1
\t\t} else {
\t\t\tlow = partitionX + 1
\t\t}
\t}

\treturn 0.0
}

func max(a, b int) int {
\tif a > b {
\t\treturn a
\t}
\treturn b
}

func min(a, b int) int {
\tif a < b {
\t\treturn a
\t}
\treturn b
}

File: binary-search/4-Median-of-Two-Sorted-Arrays.py
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        if len(nums1) > len(nums2):
            nums1, nums2 = nums2, nums1
        len1, len2 = len(nums1), len(nums2)

        low, high = 0, len1
        while low <= high:
            partitionSmallArr = (low + high) // 2
            partitionLargeArr = (len1 + len2 + 1) // 2 - partitionSmallArr

            maxLeftSmall = float('-inf') if partitionSmallArr == 0 else nums1[partitionSmallArr - 1]
            minRightSmall = float('inf') if partitionSmallArr == len1 else nums1[partitionSmallArr]
            
            maxLeftLarge = float('-inf') if partitionLargeArr == 0 else nums2[partitionLargeArr - 1]
            minRightLarge = float('inf') if partitionLargeArr == len2 else nums2[partitionLargeArr]

            if maxLeftSmall <= minRightLarge and maxLeftLarge <= minRightSmall:
                if (len1 + len2) % 2 == 0:
                    return (max(maxLeftSmall, maxLeftLarge) + min(minRightSmall, minRightLarge)) / 2
                else:
                    return max(maxLeftSmall, maxLeftLarge)
            elif maxLeftSmall > minRightLarge:
                high = partitionSmallArr - 1
            else:
                low = partitionSmallArr + 1
            
